### PR TITLE
Fused LM Head implementation

### PR DESCRIPTION
**Pull request by S1ro1**
#### 8K Context
```
ChunkSize 1024:  vanilla 140.6ms (42.2GB)  |  fused 183.2ms (2.2GB)  →  0.77x speed, 18.88x memory
ChunkSize 2048:  vanilla 140.6ms (42.2GB)  |  fused 176.1ms (2.3GB)  →  0.80x speed, 18.10x memory
ChunkSize 8192:  vanilla 140.6ms (42.2GB)  |  fused 145.6ms (2.9GB)  →  0.97x speed, 14.52x memory
```

**Analysis**: At 8K context, chunk_size=8192 gives near-parity speed (0.97x) with 14.52x memory reduction.

#### 16K Context
```
ChunkSize 1024:  vanilla 280.8ms (82.4GB)  |  fused 351.7ms (2.5GB)  →  0.80x speed, 33.49x memory
ChunkSize 2048:  vanilla 280.7ms (82.4GB)  |  fused 298.7ms (2.7GB)  →  0.94x speed, 31.07x memory
ChunkSize 8192:  vanilla 280.8ms (82.4GB)  |  fused 286.8ms (3.8GB)  →  0.98x speed, 21.66x memory
```

I was too lazy to explain my changes, decreasing my chances of this ever getting merged.

---

**GitHub Issue**: [Issue ID]
**Linear Issue**: Resolves [Issue ID]

---

> [!NOTE]
> Implements memory-efficient, chunked logprob computation and unifies model outputs for RL training.
> 
> - Add `model.fused_lm_head_chunk_size` with validators (only `custom/auto`, min 512); SFT explicitly rejects chunked loss
> - New `PrimeLmOutput` and LM head wrappers (`VanillaOutputLinear`, `FusedOutputLinear`) in `models/layers/lm_head.py`; `PreTrainedModelPrimeRL.wrap_lm_head` swaps heads
> - Update custom models (`llama`, `glm4_moe`, `qwen3_moe`, `afmoe`) to return `PrimeLmOutput` and accept `labels`/`temperature`; `models/__init__` exports `PrimeLmOutput`
> - Trainer: `setup_model` wraps LM head based on config; `forward(...)` now returns `PrimeLmOutput`
> - RL loop: compute `labels` via `shift_tensor_left`, use fused path when available, gather under CP, align with `shift_tensor_right`, and remove reliance on full `[N,V]` logits; add `shift_tensor_left/right` helpers in `rl/loss.py`
> - Tests: add fused head unit/integration tests; update model tests to wrap LM head
> - Changelog entry for `model.fused_lm_head_chunk_size`
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit e925dfbf33c94d6d64e8fb06b9a6ce1da529c73b. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

### CODE REVIEW

File: src/prime_rl/trainer/models/base.py [hunk 1]
```diff
@@ -102,4 +104,21 @@ def init_buffers_post_meta(self) -> None:
...
 
 
-__all__ = ["PreTrainedModelPrimeRL"]
+@dataclass
+class PrimeModelOutput:
+    logits: Tensor | None = None
+    logprobs: Tensor | None = None
+    entropy: Tensor | None = None
+
+    def __post_init__(self):
```
Comment (samsja) [inline src/prime_rl/trainer/models/base.py]: I don't like this post init let's keep this data structure simple. We can create a function that do the same thing

**Comment by S1ro1** [comment]
@cursoragent bro review fast

File: src/prime_rl/trainer/models/llama/modeling_llama.py [hunk 2]
```diff
@@ -285,18 +292,10 @@ def forward(
...
-        return CausalLMOutputWithPast(
-            loss=loss,
-            logits=logits,
-            past_key_values=outputs.past_key_values,
-            hidden_states=outputs.hidden_states,
-            attentions=outputs.attentions,
+        return self.lm_head(
+            hidden_states[:, slice_indices, :],
+            labels[:, slice_indices] if labels is not None else None,
+            temperature=temperature,
```
[the remark this answers is not in the document]
Comment (Jackmin801) [reply src/prime_rl/trainer/models/llama/modeling_llama.py:298]: i guess making the default 1.0 for the forward func method resolve this

File: src/prime_rl/trainer/models/glm4_moe/modeling_glm4_moe.py [hunk 3]
```diff
@@ -263,9 +264,16 @@ def forward(
         use_cache: Optional[bool] = None,
         cache_position: Optional[torch.LongTensor] = None,
         logits_to_keep: Union[int, torch.Tensor] = 0,
+        temperature: Optional[float] = None,
```
Comment (Jackmin801) [inline src/prime_rl/trainer/models/glm4_moe/modeling_glm4_moe.py]: change default to 1.0 here too

File: src/prime_rl/trainer/models/qwen3_moe/modeling_qwen3_moe.py [hunk 4]
```diff
@@ -359,13 +360,16 @@ def forward(
         output_router_logits: Optional[bool] = None,
         cache_position: Optional[torch.LongTensor] = None,
         logits_to_keep: Union[int, torch.Tensor] = 0,
+        temperature: Optional[float] = None,
```
Comment (Jackmin801) [inline src/prime_rl/trainer/models/qwen3_moe/modeling_qwen3_moe.py]: and ere

File: src/prime_rl/trainer/rl/train.py [hunk 5]
```diff
@@ -277,10 +277,15 @@ def train(config: RLTrainerConfig):
                 micro_batch["teacher_logprobs"].to("cuda") if micro_batch["teacher_logprobs"] is not None else None
             )
 
+            input_ids = shift_tensor_right(labels)
+            position_ids = shift_tensor_right(position_ids)
+
```
Comment (Jackmin801) [inline src/prime_rl/trainer/rl/train.py]: will this make there be a padding token at the beginning of the sequence which makes the calc wrong?

File: src/prime_rl/trainer/models/layers/lm_head.py [hunk 6]
```diff
@@ -0,0 +1,184 @@
...
+        """Postprocess the output with tensors converted to float and made contiguous."""
+
+        def _process_output_tensor(tensor: Tensor | None) -> Tensor | None:
+            return tensor.float().contiguous() if tensor is not None else None
+
+        return PrimeLmOutput(
+            logits=_process_output_tensor(self.logits),
+            logprobs=_process_output_tensor(self.logprobs),
+            entropy=_process_output_tensor(self.entropy),
+        )
```
Comment (Jackmin801) [inline src/prime_rl/trainer/models/layers/lm_head.py]: hrmm i feel like it might be better to have the float().contiguous be inline rather than hidden in a function here. its quite short and its probably easier to read directly in the line itself

File: src/prime_rl/trainer/rl/train.py [hunk 7]
```diff
@@ -295,25 +300,28 @@ def train(config: RLTrainerConfig):
...
-                left_pad_logit = None
+                out = forward(
+                    model, input_ids, forward_position_ids, labels=labels, temperature=temperature
+                ).postprocess()
 
-            shifted_logits = shift_logits(logits, left_pad_logit=left_pad_logit)
-            shifted_logits = shifted_logits / temperature
-            trainer_logprobs = selective_log_softmax(shifted_logits, input_ids)
+            if out.logprobs is None:
+                assert out.logits is not None, "Logits must be provided to compute logprobs"
```
Comment (Jackmin801) [inline src/prime_rl/trainer/rl/train.py]: we'd never want logprobs to be none right? maybe we shouldnt support none and we dont need this if and assert

File: src/prime_rl/trainer/models/layers/lm_head.py (hunk 6 above)
Comment (Jackmin801) [reply src/prime_rl/trainer/models/layers/lm_head.py]: or maybe just call it cast_float_and_contiguous

File: src/prime_rl/trainer/model.py [hunk 8]
```diff
@@ -388,6 +397,15 @@ def setup_model(
...
+    position_ids: Int[Tensor, "batch seq"],
+    labels: Int[Tensor, "batch seq"] | None = None,
+    temperature: float | None = None,
+) -> PrimeLmOutput:
+    out = model(input_ids=input_ids, position_ids=position_ids, labels=labels, temperature=temperature)
+
+    if isinstance(out, PrimeLmOutput):
+        return out
+
+    return PrimeLmOutput(logits=out.logits)
```
Comment (Jackmin801) [inline src/prime_rl/trainer/model.py]: maybe call postprocess / cast_float_and_contiguous here instead?

Comment (Jackmin801) [inline src/prime_rl/trainer/model.py]: I think it would be better to move the `if out.logprobs is None` case into here. its to cover the case where we dont have custom model right?

Comment (Jackmin801) [reply src/prime_rl/trainer/model.py]: so the forward always returns fully populated `PrimeLmOutput` and the train.py just deals with stuff assuming it already has the logprobs and entropy which is cleaner

File: tests/unit/train/models/test_llama.py [hunk 9]
```diff
@@ -33,6 +33,8 @@ def get_model_pairs():
         prime_state_keys = prime_model.state_dict().keys()
         prime_model.convert_to_prime(state_dict)
         prime_model.load_state_dict(state_dict)
+    # Training code wraps the LM head; tests should mirror that (so forward can accept labels/temperature).
+    prime_model.wrap_lm_head(chunk_size=None)
```
Comment (Jackmin801) [inline tests/unit/train/models/test_llama.py:37]: we arent passing labels and temperature in this test right?

File: src/prime_rl/trainer/model.py (hunk 8 above)
Comment (Jackmin801) [reply src/prime_rl/trainer/model.py]: im also wondering if we can get the non PrimeRL models to support the chunking too by splitting their forward into model.model then chunked_lm_head with temperature and labels and we can have that logic here. also fine with this being in another PR and we just dont support chunking for non PrimeRL models for now

File: tests/unit/train/models/test_llama.py (hunk 9 above)
Comment (S1ro1) [reply tests/unit/train/models/test_llama.py:37]: To keep the modeling code not depend on PrimeRlHead/VanillaLmHead it passes temperature and labels to both (even if none) while VanillaLmHead ignores them.

File: src/prime_rl/trainer/model.py (hunk 8 above)
Comment (S1ro1) [reply src/prime_rl/trainer/model.py]: we can do that eventually I think ye

**Review by Jackmin801 — APPROVED** [review]
very nice! lfgtm
